runtime/sema/check_composite_declaration.go

```diff
@@ -1937,29 +2109,42 @@ func (checker *Checker) checkCompositeFunctions(
 	}
 }
 
-func (checker *Checker) declareSelfValue(selfType Type, selfDocString string) {
-
-	// NOTE: declare `self` one depth lower ("inside" function),
+func (checker *Checker) declareLowerScopedValue(
```

Comment:
Good idea to generalize this!

Maybe add a comment to this function explaining what "lower" means (inside), and provide example use cases (e.g. for self and base)